### DISCUSSION TITLE
Unable to install torch_sparse==0.6.7 with correct CUDA version

### DISCUSSION THREAD

**Discussion by LucaMatarazzo** [pyg-team/pytorch_geometric#8850] [Q&A] [answered]
## Problem

Hi all, I am trying to install `torch_sparse==0.6.7` in a virtual environment on a linux server to replicate the [following paper](https://arxiv.org/abs/2106.02172). I can install `torch_sparse==0.6.7` correctly in the venv but when trying to import it I get the following error: 

`Detected that PyTorch and torch_sparse were compiled with different CUDA versions. PyTorch has CUDA version 10.2 and torch_sparse has CUDA version 11.4. Please reinstall the torch_sparse that matches your PyTorch install.`

## Tried solutions

To solve it, I uninstalled it (`pip uninstall torch_sparse`) and tried the following solutions:

1. `pip install torch-sparse==0.6.7+cu102 -f https://pytorch-geometric.com/whl/torch-1.6.0.html` (as originally suggested [here](https://github.com/rusty1s/pytorch_sparse/issues/49)). But I get the following error: 

`ERROR: Could not find a version that satisfies the requirement torch-sparse==0.6.7+cu102 (from versions: latest+cpu, latest+cu92, latest+cu101, latest+cu102, 0.1.0, 0.2.0, 0.2.1, 0.2.2, 0.2.3, 0.2.4, 0.4.0, 0.4.1, 0.4.2, 0.4.3, 0.4.4, 0.5.1, 0.6.0, 0.6.1, 0.6.3, 0.6.4, 0.6.5, 0.6.6, 0.6.7, 0.6.7+cpu, 0.6.7+cu101, 0.6.7+cu102, 0.6.7+cu92, 0.6.8, 0.6.8+cpu, 0.6.8+cu101, 0.6.8+cu102, 0.6.8+cu92, 0.6.9, 0.6.10, 0.6.11, 0.6.12)
ERROR: No matching distribution found for torch-sparse==0.6.7+cu102`

2. `pip install torch-scatter torch-sparse -f https://data.pyg.org/whl/torch-1.6.0+cu102.html` (as indicated [here](https://github.com/rusty1s/pytorch_sparse?tab=readme-ov-file)). But I get the following error:

`Looking in links: https://data.pyg.org/whl/torch-1.6.0+cu102.html
Collecting torch-scatter
  Using cached torch_scatter-2.0.9.tar.gz (21 kB)
  Preparing metadata (setup.py) ... done
Collecting torch-sparse
  Using cached torch_sparse-0.6.12.tar.gz (43 kB)
  Preparing metadata (setup.py) ... done
Requirement already satisfied: scipy in ./.venv/lib/python3.6/site-packages (from torch-sparse) (1.5.4)
Requirement already satisfied: numpy>=1.14.5 in ./.venv/lib/python3.6/site-packages (from scipy->torch-sparse) (1.19.2)
Building wheels for collected packages: torch-scatter, torch-sparse
  Building wheel for torch-scatter (setup.py) ... error
  ERROR: Command errored out with exit status 1:`
[...]
`  ERROR: Failed building wheel for torch-scatter
  Running setup.py clean for torch-scatter
  Building wheel for torch-sparse (setup.py) ... error
  ERROR: Command errored out with exit status 1:`
[...]
`  ERROR: Failed building wheel for torch-sparse
  Running setup.py clean for torch-sparse
Failed to build torch-scatter torch-sparse
Installing collected packages: torch-sparse, torch-scatter
    Running setup.py install for torch-sparse ... error
    ERROR: Command errored out with exit status 1:`
[...]
`    error: command 'gcc' failed with exit status 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /home/myname/directoryname/.venv/bin/python3 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-uu1cqap8/torch-sparse_908976f98b4e496880675245c7dab3f0/setup.py'"'"'; __file__='"'"'/tmp/pip-install-uu1cqap8/torch-sparse_908976f98b4e496880675245c7dab3f0/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-mwnswc97/install-record.txt --single-version-externally-managed --compile --install-headers /home/matarazzo/CFLP/.venv/include/site/python3.6/torch-sparse`

## Environment

I am using Python 3.6.8 (while the authors originally use 3.8.5), pip 21.3.1 and installed in the venv the following:

- torch_geometric==1.6.1      
- torch==1.6.0                
- matplotlib==3.3.1
- scikit_network==0.15.0      
- numpy==1.19.2               
- six==1.15.0    
- networkx==2.5
- dgl==0.5.1                  
- scipy==1.5.4                
- geomloss==0.2.4
- ogb==1.3.1
- scikit_learn==0.24.2
- wheel==0.37.1

**Comment by rusty1s** [comment]
Try
```
pip install torch-scatter torch-sparse --no-index -f https://data.pyg.org/whl/torch-1.6.0+cu102.html

**Reply by LucaMatarazzo** [reply]
Hi, thanks a lot for the advice. The installation now works, but I get another error when importing torch_sparse:

`Traceback (most recent call last):
  File "main.py", line 6, in <module>
    import torch_sparse
  File "/home/myname/directory/.venv/lib64/python3.6/site-packages/torch_sparse/__init__.py", line 15, in <module>
    f'{library}_{suffix}', [osp.dirname(__file__)]).origin)
  File "/home/myname/directory/.venv/lib64/python3.6/site-packages/torch/_ops.py", line 105, in load_library
    ctypes.CDLL(path)
  File "/usr/lib64/python3.6/ctypes/__init__.py", line 343, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libcusparse.so.10: cannot open shared object file: No such file or directory`

Again, didn't manage to find an appropriate response online :(

**Reply by rusty1s** [reply]
That would mean `torch-sparse` cannot find the `libcusparse` dependency on your machine. In later versions of `torch-sparse`, `libcusparse` is no longer a dependency. For older versions, you might have to specify the path to `libcusparse` (coming from your CUDA installation) in `LD_LIBRARY_PATH` (usually something like `{path_to_cuda}/lib`).

**Reply by LucaMatarazzo** [reply]
Hi, thanks for the reply. My path to CUDA should be `/usr/local/cuda`. Inside, there is a `lib64`. `LD_LIBRARY_PATH` was empty before, and I defined as follows: `export LD_LIBRARY_PATH=/usr/local/cuda/lib64:`. Still, when running the script I run in the same error as before importing `torch_sparse`. I noticed that listing the contents of `/usr/local/cuda/lib64`, among others, I find inside the following:

> libcusparse.so
> libcusparse.so.11
> libcusparse.so.11.6.0.120
> libcusparse_static.a

However, there is no libcusparse.so.10. I'm not sure if that's the problem and how to solve it...

EDIT: also, running `print(torch.version.cuda)` I obtain 10.2, as expected, while running `nvcc --version` in the terminal I obtain:

> nvcc: NVIDIA (R) Cuda compiler driver
> Copyright (c) 2005-2021 NVIDIA Corporation
> Built on Mon_Oct_11_21:27:02_PDT_2021
> Cuda compilation tools, release 11.4, V11.4.152
> Build cuda_11.4.r11.4/compiler.30521435_0

which suggests I have maybe multiple CUDA installations. Again, not sure on how to set things properly while keeping the correct versions of `torch` and `torch_geometric` (1.6.0 and 1.6.1). Note that I installed using `pip install torch==1.6.0`.

**Reply by rusty1s** [reply]
You can take a look in your installation folder of PyTorch, I would assume `libcusparse.so.10` is included in there.

**Reply by LucaMatarazzo** [reply]
Unfortunately it seems that's not the case: with `pip show torch` I get:

> Location: /home/myname/directoryname/.venv/lib/python3.6/site-packages

However, using `find .venv -name "libcusparse*"` in the terminal to look for the file inside the venv where `torch` was installed, I get nothing. I don't know if that's useful, but using `find .venv -name "libc*"`, I get the following:

> .venv/lib/python3.6/site-packages/torch/lib/libc10_cuda.so
> .venv/lib/python3.6/site-packages/torch/lib/libcaffe2_detectron_ops_gpu.so
> .venv/lib/python3.6/site-packages/torch/lib/libcaffe2_observers.so
> .venv/lib/python3.6/site-packages/torch/lib/libc10.so
> .venv/lib/python3.6/site-packages/torch/lib/libcaffe2_nvrtc.so
> .venv/lib/python3.6/site-packages/torch/lib/libcudart-80664282.so.10.2
> .venv/lib/python3.6/site-packages/torch/lib/libcaffe2_module_test_dynamic.so

**Reply by rusty1s** [reply, marked as answer]
Did you install torch via miniconda? You may also find it in `miniconda/lib` path.

**Reply by LucaMatarazzo** [reply]
EDIT: After some more trials, it worked. Thank you for the fast and useful support! For others with similar issues: I used `find / -name "libcusparse.so.10" -print 2>/dev/null` to search in the whole system for `libcusparse.so.10` and found this path:

> /usr/local/tmp/pisa/.conda/pkgs/cudatoolkit-10.2.89-hfd86e86_1/lib/libcusparse.so.10

updated `LD_LIBRARY_PATH` as follows: `export LD_LIBRARY_PATH=/usr/local/tmp/pisa/.conda/pkgs/cudatoolkit-10.2.89-hfd86e86_1/lib:` and then was able to import `torch_sparse`. 

Unfortunately I ran in another issue with CUDA but that's for another discussion forum. Thanks again! :)
